File: native-plugins/admob/resources/ios/Sources/BannerLifecycle.swift
import Foundation
import UIKit

/// Wires AdMob banner lifecycle into NativePHP's UIScene notifications.
///
/// Unlike Android, iOS GADBannerView does not require explicit
/// resume/pause calls - the system handles that. We still register the
/// observers so that any future per-banner cleanup (e.g. clearing on
/// background termination) has a known hook point.
///
/// Registered once after the explicit `Admob.Initialize` bridge call.
enum BannerLifecycle {
    private static var registered = false

    static func register() {
        guard !registered else { return }
        registered = true

        NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { _ in
            // Banners auto-refresh; nothing required here. Future per-banner
            // hooks can iterate BannerRegistry.shared.all().
        }

        NotificationCenter.default.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: .main
        ) { _ in
            // No-op for now; intentionally a hook point.
        }

        NotificationCenter.default.addObserver(
            forName: UIApplication.willTerminateNotification,
            object: nil,
            queue: .main
        ) { _ in
            BannerRegistry.shared.clear()
        }
    }
}

File: native-plugins/admob/resources/android/src/BannerLifecycle.kt
package com.blessedzulu.nativephp.admob

import com.nativephp.mobile.lifecycle.NativePHPLifecycle

/**
 * Wires the AdMob banner lifecycle into NativePHP's activity callbacks.
 *
 * Google's AdView SDK requires resume() / pause() at the right moments to
 * keep impressions accurate and to avoid background battery drain.
 * onDestroy() releases the native resources.
 *
 * Registered once after the explicit Admob.Initialize bridge call.
 */
object BannerLifecycle {
    private var registered = false

    @JvmStatic
    @Synchronized
    fun register() {
        if (registered) return
        registered = true

        NativePHPLifecycle.on("onResume") {
            BannerRegistry.all().forEach { it.resume() }
        }

        NativePHPLifecycle.on("onPause") {
            BannerRegistry.all().forEach { it.pause() }
        }

        NativePHPLifecycle.on("onDestroy") {
            BannerRegistry.clear()
        }
    }
}
